src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/resource/AbstractResourcesProvider.kt

```diff
@@ -10,45 +10,70 @@
  ******************************************************************************/
 package org.jboss.tools.intellij.kubernetes.model.resource
 
+import com.intellij.openapi.diagnostic.logger
 import io.fabric8.kubernetes.api.model.HasMetadata
-import org.jboss.tools.intellij.kubernetes.model.util.areEqual
+import org.jboss.tools.intellij.kubernetes.model.util.sameUid
 
 abstract class AbstractResourcesProvider<R : HasMetadata> : IResourcesProvider<R> {
 
-    protected val allResources = mutableSetOf<R>()
+    protected val allResources = mutableListOf<R>()
 
     override fun invalidate() {
+        logger<AbstractResourcesProvider<*>>().debug("Invalidating all $kind resources.")
         synchronized(allResources) {
             allResources.clear()
         }
     }
 
-    override fun invalidate(resource: HasMetadata) {
-        synchronized(allResources) {
-            allResources.remove(resource)
-        }
-    }
-
     override fun add(resource: HasMetadata): Boolean {
-        if (!kind.clazz.isAssignableFrom(resource::class.java)) {
+        if (!isOfCorrectKind(resource)) {
             return false
         }
+        logger<AbstractResourcesProvider<*>>().debug("Adding resource ${resource.metadata.name}.")
         // don't add resource if different instance of same resource is already contained
         synchronized(allResources) {
-            return allResources.find { areEqual(it, resource) } == null
-                    && allResources.add(resource as R)
+            val existing = allResources.find { resource.sameUid(it) }
+            return when (existing) {
+                resource -> false
+                null -> allResources.add(resource as R)
+                else -> replaceBy(existing, resource)
```

Comment:
Why do you keep this else? When does `find` return a different value than a `resource` or `null`?

Reply:
my understanding is the following:
* resource: it has found the very same instance, no need to replace it
* null: no object with the same uid was found
* else: a different instance with the same uid was found, we have to replace it. (ex. the watch has an update for a given object)

Maybe the case `resource` is very unlikely and could be dropped? Missing something?

Reply:
No, i agree with your solution as well. It makes completely sense.